Fix open Contact tab from "customers only" message

duleorlovic opened the pull request:
#### What? Why?

We need stimulus attributes that will activate Contact tab

- Closes #11784 

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->



#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- Got to enterprise settings, eg local link is http://localhost:3000/admin/enterprises/fred-s-farm/edit#/shop_preferences_panel
- Check " VISIBLE TO REGISTERED CUSTOMERS ONLY"
- Open shopfront eg local link is http://localhost:3000/fredo-s-farm-hub/shop
- Click on Contact link

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

rioug reviewed (approved):
Nice one ! Thanks for your help.
I noticed you merged master into your branch, we prefer rebasing the dev branch on master to keep the git history nice an clean.

dacook reviewed (approved):
Thank you!

That code seems quite wordy, but I see that it's the same method as used elsewhere. Hopefully one day we might be able to simplify it..

duleorlovic commented:
> Nice one ! Thanks for your help.
> I noticed you merged master into your branch, we prefer rebasing the dev branch on master to keep the git history nice an clean.

Sorry, I have clicked on the github page that I want to rebase on master... Now I have rebased locally and force pushed, so now there is only one commit

duleorlovic commented:
> Thank you!
> 
> That code seems quite wordy, but I see that it's the same method as used elsewhere. Hopefully one day we might be able to simplify it..

@dacook you are right. We should not wait for refactoring. I added a another commit that solves this [comment](https://github.com/openfoodfoundation/openfoodnetwork/issues/11784#issuecomment-1809544095) , mainly we do not need to use two methods, just `data-action="tabs-and-panels#activate"`.

Removed shop-tabs controllers since we can listen on `"data-action": "orderCycleSelected@window->tabs-and-panels#activateDefaultPanel"`
    
 Added test for cases:
    * activate by clicking on tab
    * activateDefaultPanel on orderCycleSelected event
    * activateFromWindowLocationOrDefaultPanelTarget to activate tab based on achor in URL

@dacook @rioug Please review again

rioug commented on spec/javascripts/stimulus/tabs_and_panels_controller_test.js:
```diff
@@ -11,122 +11,143 @@ describe('TabsAndPanelsController', () => {
+  describe("when non valid anchor is specified in the url", () => {
+    const oldWindowLocation = window.location
+    beforeAll(() => {
+      Object.defineProperty(window, "location", {
+        value: new URL("http://example.com/#non_valid_panel"),
+        configurable: true,
```
It's a lot cleaner than the previous solution !

duleorlovic replied:
Yea, no need to use mock

rioug reviewed (approved):
Looks great ! thanks for cleaning that up :pray:

mkllnk commented:
There are a couple of specs failing.

sigmundpetersen commented:
@duleorlovic did you want to fix the failing specs?
Then this can move forward 👍

duleorlovic commented:
> There are a couple of specs failing.

Sorry for late reply. I was thinking that there were som Flaky test, but I had to update some test case also.
I used `data-test=link_for_enterprise_fees` instead of `id=enterprise_fees` so it is obvious that we are using for `expect(page).to have_selector "[data-test=link_for_enterprise_fees]"`

mkllnk reviewed (approved):
Great, thank you!

drummer83 commented:
Hi @duleorlovic,
Thank you for this contribution! I have tested this PR on our staging server and it seems to work great! :muscle: 

I have tested that
- with the 'registered users only' option **enabled**
  - guest users see the contact message and clicking the contact link works well, as well as the login link,
  - logged-in users (no customers yet) see the contact message and clicking the link works well,
  - logged-in customers see the products right away,
- with the 'registered users only' option **disabled**
  - guest users see the products right away,
  - logged-in users (no customers yet) see the products right away,
  - logged-in customers see the products right away.

Here is a short video (covering only parts of the tests):

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/1b173c5d-4f0d-406b-83a1-cd833a17b266

I think this one ready to go! :partying_face: 
I'll go ahead and merge this one! :rocket: 
Thanks again!
